src/MMALSharp.Processing/Processors/Motion/FrameDiffDriver.cs

```diff
@@ -0,0 +1,311 @@
+﻿// <copyright file="FrameDiffDriver.cs" company="Techyian">
+// Copyright (c) Ian Auty and contributors. All rights reserved.
+// Licensed under the MIT License. Please see LICENSE.txt for License info.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Drawing;
+using System.Drawing.Imaging;
+using System.IO;
+using System.Runtime.InteropServices;
+using Microsoft.Extensions.Logging;
+using MMALSharp.Common;
+using MMALSharp.Common.Utility;
+
+namespace MMALSharp.Processors.Motion
+{
+    /// <summary>
+    /// A frame difference motion detection base class which buffers a test frame and a current frame,
+    /// stores frame metrics, and invokes an <see cref="IMotionAlgorithm"/> to analyse full frames.
+    /// </summary>
+    public class FrameDiffDriver : FrameAnalyser
+    {
+        // Prefer fields over properties for parallel processing performance reasons.
+        // Parallel processing references unique array indices, so arrays do not need
+        // to be stored in the passed-by-value FrameDiffMetrics struct.
+
+        // Various frame properties are collected when the first full frame is available.
+        // A by-val copy of this struct is passed into the parallel processing algorithm. It is
+        // private rather than internal to prevent accidental parallel reads against this copy.
+        private FrameDiffMetrics _frameMetrics;
+
+        private Action _onDetect;
+        private bool _firstFrame = true;
+        private MotionConfig _motionConfig;
+        private bool _fullTestFrame;
+        private ImageContext _imageContext;
+        private Stopwatch _testFrameAge;
+
+        /// <summary>
+        /// Fully are skipped when comparing the test frame to the current frame.
```

Comment:
Think this is a typo? Is it supposed to read "Frames" instead of "Fully"?

Reply:
It should say "Fully black pixels", not sure how that happened. It is fixed in the changes I haven't committed yet (more about that momentarily).